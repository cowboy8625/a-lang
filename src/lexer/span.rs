#![allow(unused)]
#[derive(Debug, Default, Clone, Copy, Hash, PartialEq, Eq, PartialOrd, Ord)]
pub struct Span {
    pub start: usize,
    pub end: usize,
    pub line: usize,
}

impl Span {
    pub fn new(line: usize, start: usize, end: usize) -> Self {
        Self { start, end, line }
    }
    pub fn right_shift(&mut self, ch: char) {
        if ch == '\0' {
            return;
        }
        if ch == '\n' {
            self.line += 1;
        }
        self.end += ch.to_string().as_bytes().len();
    }

    pub fn range(&self) -> std::ops::Range<usize> {
        self.start..self.end
    }
}

impl From<&Span> for Span {
    fn from(other: &Self) -> Self {
        Self {
            start: other.end,
            end: other.end,
            line: other.line,
        }
    }
}

impl From<(Span, Span)> for Span {
    fn from((x, y): (Self, Self)) -> Self {
        Self {
            start: x.start,
            end: y.end,
            line: x.line,
        }
    }
}
