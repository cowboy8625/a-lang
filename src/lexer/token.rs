use super::{Span, TokenKind};

macro_rules! is_a {
    ($kind:ident, $id:ident, $name:ident) => {
        pub fn $name(&self) -> bool {
            match self {
                $kind::$id(..) => true,
                _ => false,
            }
        }
    };
}

#[derive(Debug, Clone)]
pub enum Token {
    KeyWord(String, Span),
    Id(String, Span),
    Int(String, Span),
    Op(String, Span),
    Ctrl(String, Span),
    String(String, Span),
    Char(char, Span),
    Error(String, Span),
    Eof(Span),
}

impl Token {
    pub fn kind(&self) -> TokenKind {
        match &self {
            Self::KeyWord(ref word, ..) => match word.as_str() {
                "true" => TokenKind::True,
                "false" => TokenKind::False,
                "use" => TokenKind::Use,
                "if" => TokenKind::If,
                "else" => TokenKind::Else,
                "struct" => TokenKind::Struct,
                "fn" => TokenKind::Fn,
                "return" => TokenKind::Return,
                "let" => TokenKind::Let,
                _ => unreachable!(),
            },
            Self::Id(..) => TokenKind::Id,
            Self::Int(..) => TokenKind::Int,
            Self::Ctrl(ctrl, ..) => match ctrl.as_str() {
                ";" => TokenKind::SemiColon,
                ":" => TokenKind::Colon,
                "(" => TokenKind::LParan,
                ")" => TokenKind::RParan,
                "[" => TokenKind::LBracet,
                "]" => TokenKind::RBracet,
                "{" => TokenKind::LBrace,
                "}" => TokenKind::RBrace,
                "->" => TokenKind::RightArrow,
                "=>" => TokenKind::ThickRightArrow,
                "," => TokenKind::Comma,
                _ => unreachable!(),
            },
            Self::Op(op, ..) => match op.as_str() {
                "+" => TokenKind::Plus,
                "+=" => TokenKind::PlusEq,
                "-" => TokenKind::Minus,
                "*" => TokenKind::Star,
                "/" => TokenKind::Slash,
                "==" => TokenKind::EqEq,
                "=" => TokenKind::Eq,
                "!=" => TokenKind::Neq,
                ">=" => TokenKind::Geq,
                "<=" => TokenKind::Leq,
                "<" => TokenKind::Les,
                ">" => TokenKind::Grt,
                "!" => TokenKind::Not,
                "|" => TokenKind::Pipe,
                "." => TokenKind::Dot,
                _ => unimplemented!(),
            },
            Self::String(..) => TokenKind::String,
            Self::Char(..) => TokenKind::Char,
            Self::Eof(..) => TokenKind::Eof,
            _ => unimplemented!(),
        }
    }
    pub fn lexme(&self) -> String {
        match &self {
            Self::KeyWord(i, ..) => i.to_string(),
            Self::Id(i, ..) => i.to_string(),
            Self::Int(i, ..) => i.to_string(),
            Self::Ctrl(i, ..) => i.to_string(),
            Self::Op(i, ..) => i.to_string(),
            Self::String(i, ..) => i.to_string(),
            Self::Char(i, ..) => i.to_string(),
            Self::Error(i, ..) => i.to_string(),
            Self::Eof(..) => "Eof".to_string(),
        }
    }

    pub fn span(&self) -> Span {
        match &self {
            Self::KeyWord(.., span) => *span,
            Self::Int(.., span) => *span,
            Self::Id(.., span) => *span,
            Self::Ctrl(.., span) => *span,
            Self::Op(.., span) => *span,
            Self::String(.., span) => *span,
            Self::Char(.., span) => *span,
            Self::Eof(span) => *span,
            _ => unreachable!(),
        }
    }
    pub fn is_keyword_of(&self, expect: &str) -> bool {
        match self {
            Token::KeyWord(word, ..) if expect == word => true,
            _ => false,
        }
    }
    pub fn is_op_of(&self, expect: &str) -> bool {
        match self {
            Token::Op(op, ..) if expect == op => true,
            _ => false,
        }
    }

    pub fn is_err(&self) -> bool {
        match self {
            Token::Error(..) => true,
            _ => false,
        }
    }

    is_a!(Token, KeyWord, is_keyword);
    is_a!(Token, Int, is_int);
    is_a!(Token, Id, is_id);
    is_a!(Token, Op, is_op);
    is_a!(Token, Eof, is_eof);
    pub fn lookup(ident: &str, span: Span) -> Option<Self> {
        match ident {
            "fn" | "struct" | "if" | "else" | "use" | "return" | "let" | "true" | "false" => {
                Some(Self::KeyWord(ident.into(), span))
            }
            _ => None,
        }
    }
}
