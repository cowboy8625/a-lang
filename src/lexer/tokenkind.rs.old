#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum TokenKind {
    Int,
    String,
    Char,
    Id,
    Use,
    Let,
    Struct,
    True,
    False,
    Fn,
    If,
    Else,
    Return,
    Plus,
    Minus,
    Star,            // *
    Slash,           // /
    SemiColon,       // ;
    Colon,           // :
    Comma,           // ,
    Dot,             // .
    LBrace,          // {
    RBrace,          // }
    LBracet,         // [
    RBracet,         // ]
    LParan,          // (
    RParan,          // )
    RightArrow,      // ->
    ThickRightArrow, // =>
    Grt,             // <
    Les,             // >
    Geq,             // >=
    Leq,             // <=
    Neq,             // !=
    Not,             // !
    Eq,              // =
    EqEq,            // ==
    PlusEq,          //  +=,
    Pipe,            //  |
    Eof,
}
