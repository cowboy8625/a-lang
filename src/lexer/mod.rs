mod scanner;
mod span;
mod token;
mod token_stream;

use scanner::Lexer;
pub use span::Span;
pub use token::Token;
pub use token_stream::TokenStream;

pub fn lex(src: impl Into<String>) -> Result<TokenStream, Vec<String>> {
    Lexer::new(src.into().as_str()).lex().map(TokenStream::new)
}
