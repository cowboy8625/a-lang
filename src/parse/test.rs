use pretty_assertions::assert_eq;
use super::*;
use crate::lexer::lex;

pub fn snapshot_lexing(input: &str) -> String {
    let tokens = match lex(input) {
        Ok(l) => l,
        Err(e) => return e.into_iter().map(|i| format!("{i}\n")).collect(),
    };
    let mut tokens = std::collections::VecDeque::from(tokens.stream);
    let mut output = String::new();
    let mut idx = 0;
    for (row, line) in input.lines().enumerate() {
        output += line;
        output += "\n";
        let mut line_count = 0;
        while let Some(tok) = tokens.pop_front() {
            // if tok.span().start != tok.span().end {
            //     panic!("We haven't handled this yet");
            // }

            if tok.span().line > idx && tok.span().line != row {
                tokens.push_front(tok);
                break;
            }
            output += &" ".repeat(tok.span().start - idx);
            output += &"^".repeat(tok.span().end - tok.span().start);
            output += &format!(" {tok:?}");
            line_count += tok.span().len();
            output += "\n"
        }
        idx += line_count;
    }

    output
}

macro_rules! snapshot {
    ($name:tt, $path:tt) => {
        #[test]
        fn $name() {
            let contents = include_str!($path);
            let mut settings = insta::Settings::clone_current();
            settings.set_snapshot_path("testdata/output/");
            settings.bind(|| {
                insta::assert_snapshot!(snapshot_lexing(contents));
            });
        }
    };
}

snapshot!(test_lexer, "testdata/snapshots/lexer.a");

// #[test]
// fn test_primary() {
//     let tokens = lex("1").unwrap();
//     let ast = parse(tokens).unwrap();
//     insta::assert_yaml_snapshot!(ast);
// }
