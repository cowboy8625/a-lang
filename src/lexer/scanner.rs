#![allow(unused)]
use super::{Span, Token};
use crate::parse::{
    keyword,
    CtrlColon,
    CtrlComma,
    CtrlDot,
    CtrlLBrace,
    CtrlLBracet,
    CtrlLParan,
    CtrlRBrace,
    CtrlRBracet,
    CtrlRParan,
    CtrlRightArrow,
    // CtrlStar,
    // CtrlSlash,
    CtrlSemiColon,
    Ident,
    LitBool,
    LitChar,
    LitInt,
    LitStr,
    OpAdd,
    OpDiv,
    OpEqual,
    OpEqualEqual,
    OpGeq,
    OpGrt,
    OpLeq,
    OpLes,
    OpMul,
    OpNeq,
    OpNot,
    OpSub,
};
use std::iter::Peekable;
use std::str::Chars;

pub struct Lexer<'a> {
    src: Peekable<Chars<'a>>,
    len: usize,
    span: Span,
    last_chr_len: usize,
}

impl<'a> Lexer<'a> {
    pub fn new(src: &'a str) -> Self {
        Self {
            len: src.len(),
            src: src.chars().peekable(),
            span: Span::default(),
            last_chr_len: 0,
        }
    }

    fn tp(&self) -> usize {
        self.span.end
    }

    fn is_end(&mut self) -> bool {
        self.src.peek().is_none()
        // self.tp() >= self.len.saturating_sub(1)
    }

    fn peek(&mut self) -> char {
        self.src.peek().cloned().unwrap_or('\0')
    }

    fn next(&mut self) -> char {
        let ch = self.src.next().unwrap_or('\0');
        self.span.right_shift(ch);
        self.last_chr_len = ch.to_string().as_bytes().len();
        ch
    }

    fn next_if<F: FnOnce(char) -> bool>(&mut self, func: F) -> Option<char> {
        let c = self.peek();
        if func(c) {
            assert_eq!(c, self.next());
            return Some(c);
        }
        None
    }

    fn span(&mut self) -> Span {
        let span = self.span;
        self.span.start = self.span.end;
        span
    }

    fn number(&mut self, c: char) -> Box<dyn Token> {
        let mut number = c.to_string();
        while let Some(c) = self.next_if(|c| c.is_ascii_digit() || c == '_') {
            number.push(c);
        }
        Box::new(LitInt::new(number, self.span()))
    }

    fn ident(&mut self, c: char) -> Box<dyn Token> {
        let mut id = c.to_string();
        while let Some(c) = self.next_if(|c| c.is_ascii_alphanumeric() || c == '_') {
            id.push(c);
        }
        let span = self.span();
        match id.as_str() {
            "fn" => Box::new(keyword::Fn(span)),
            "struct" => Box::new(keyword::Struct(span)),
            "if" => Box::new(keyword::If(span)),
            "else" => Box::new(keyword::Else(span)),
            "use" => Box::new(keyword::Use(span)),
            "return" => Box::new(keyword::Return(span)),
            "let" => Box::new(keyword::Let(span)),
            "true" => Box::new(LitBool::new(id, span)),
            "false" => Box::new(LitBool::new(id, span)),
            _ => Box::new(Ident::new(id, span)),
        }
    }

    fn string(&mut self) -> Box<dyn Token> {
        let mut string = String::new();
        while let Some(c) = self.next_if(|c| c != '"') {
            string.push(c);
        }
        self.next();
        Box::new(LitStr::new(string, self.span()))
    }

    fn chr(&mut self) -> Box<dyn Token> {
        let mut string = String::new();
        while let Some(c) = self.next_if(|c| c != '\'') {
            string.push(c);
        }
        self.next();

        Box::new(LitChar::new(string, self.span()))
    }

    fn comment(&mut self) -> Box<dyn Token> {
        while let Some(_) = self.next_if(|c| c != '\n') {}
        let ch = self.next();
        self.parse(ch)
    }

    fn token<T>(&mut self, op: &str) -> Box<dyn Token>
    where
        T: Token,
    {
        for _ in 0..op.chars().count().saturating_sub(self.last_chr_len) {
            self.next();
        }
        Box::new(T::new(op.into(), self.span()))
    }

    fn parse(&mut self, ch: char) -> Box<dyn Token> {
        match ch {
            n @ '0'..='9' => self.number(n),
            i @ ('a'..='z' | 'A'..='Z') => self.ident(i),
            '"' => self.string(),
            '\'' => self.chr(),
            '/' if self.peek() == '/' => self.comment(),
            '-' if self.peek() == '>' => self.token::<CtrlRightArrow>("->"),
            '>' if self.peek() == '=' => self.token::<OpGeq>(">="),
            '<' if self.peek() == '=' => self.token::<OpLeq>("<="),
            '=' if self.peek() == '=' => self.token::<OpEqualEqual>("=="),
            // '|' if self.peek() == '|' => self.token::<>("||"),
            // '&' if self.peek() == '&' => self.token::<>("&&"),
            '!' if self.peek() == '=' => self.token::<OpNeq>("!="),
            // // '|' => self.op_token("|"),
            // // '&' => self.op_token("&"),
            '-' => self.token::<OpSub>("-"),
            '+' => self.token::<OpAdd>("+"),
            '*' => self.token::<OpMul>("*"),
            '/' => self.token::<OpDiv>("/"),
            '>' => self.token::<OpGrt>(">"),
            '<' => self.token::<OpLes>("<"),
            '=' => self.token::<OpEqual>("="),
            '!' => self.token::<OpNot>("!"),
            // '%' => self.op_token("%"),
            '.' => self.token::<CtrlDot>("."),
            ',' => self.token::<CtrlComma>(","),
            '(' => self.token::<CtrlLParan>("("),
            ')' => self.token::<CtrlRParan>(")"),
            '{' => self.token::<CtrlLBrace>("{"),
            '}' => self.token::<CtrlRBrace>("}"),
            '[' => self.token::<CtrlLBracet>("{"),
            ']' => self.token::<CtrlRBracet>("}"),
            ':' => self.token::<CtrlColon>(":"),
            ';' => self.token::<CtrlSemiColon>(";"),
            // 'λ' => self.op_token("λ"),
            '\n' | ' ' | '\0' => {
                let ch = self.next();
                self.span.start = self.span.end.saturating_sub(self.last_chr_len);
                self.parse(ch)
            }
            _ => panic!(),
            // '\0' => Token::Eof(self.span()),
            // c => Token::Error(format!("unknown char '{}'", c), self.span()),
        }
    }

    pub fn lex(mut self) -> Result<Vec<Box<dyn Token>>, Vec<String>> {
        let mut tokens = vec![];
        // let mut errors = vec![];
        while !self.is_end() {
            let ch = self.next();
            let token = self.parse(ch);
            tokens.push(token);
            // if !token.is_err() {
            //     tokens.push(token);
            // } else {
            //     errors.push(token.lexme().to_string())
            // }
        }
        // if !errors.is_empty() {
        //     return Err(errors);
        // }
        Ok(tokens)
    }
}
