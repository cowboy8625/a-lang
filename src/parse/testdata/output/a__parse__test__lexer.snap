---
source: src/parse/test.rs
expression: snapshot_lexing(contents)
---
fn main() {
  1 + 3;
}

